File: Srednje skole/2026/Jacina sifre/solutions/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cctype>

// Define number of rules
#define NUMBER_OF_RULES 7
#define RULE_1_LENGTH 11

// Boolean array to track which rules are activated
bool rules_activated[NUMBER_OF_RULES];

// Helper function to reset the rules array
void reset_rules_array() {
    for (int i = 0; i < NUMBER_OF_RULES; i++) {
        rules_activated[i] = false;
    }
}

// Rule 1: Minimum length (at least 8 characters)
bool rule1(const std::string& s) {
    return s.length() >= RULE_1_LENGTH;
}

// Rule 2: First character is lowercase letter (a-z)
bool rule2(const std::string& s) {
    if (s.empty()) return false;
    return std::islower(s[0]);
}

// Rule 3: Last character is a digit (0-9)
bool rule3(const std::string& s) {
    if (s.empty()) return false;
    return std::isdigit(s[s.length() - 1]);
}

// Rule 4: Either second or second-to-last character is a special character
bool rule4(const std::string& s) {
    if (s.length() < 2) return false;

    // Check second character
    char second = s[1];
    if (second == '!' || second == '@' || second == '#' || second == '$' ||
        second == '%' || second == '&' || second == '*' || second == '?' ||
        second == '_') {
        return true;
    }

    // Check second-to-last character
    char second_last = s[s.length() - 2];
    if (second_last == '!' || second_last == '@' || second_last == '#' || second_last == '$' ||
        second_last == '%' || second_last == '&' || second_last == '*' || second_last == '?' ||
        second_last == '_') {
        return true;
    }

    return false;
}

// Rule 5: At least 3 of 4 character categories
bool rule5(const std::string& s) {
    bool has_lower = false;
    bool has_upper = false;
    bool has_digit = false;
    bool has_special = false;

    for (char c : s) {
        if (c >= 'a' && c <= 'e') {
            has_lower = true;
        } else if (c >= 'A' && c <= 'E') {
            has_upper = true;
        } else if (c >= '0' && c <= '4') {
            has_digit = true;
        } else if (c == '!' || c == '@' || c == '#') {
            has_special = true;
        }
    }

    int count = (has_lower ? 1 : 0) + (has_upper ? 1 : 0) + (has_digit ? 1 : 0) + (has_special ? 1 : 0);
    return count >= 3;
}

// Rule 6: At least one character appears 4 or more times
bool rule6(const std::string& s) {
    std::vector<int> count(256, 0); // Assuming ASCII characters

    for (char c : s) {
        count[c]++;
        if (count[c] >= 4) {
            return true;
        }
    }

    return false;
}

// Rule 7: Repeated bigram (substring of length 2)
bool rule7(const std::string& s) {
    if (s.length() < 4) return false;

    for (size_t i = 0; i < s.length() - 3; i++) {
        char first_char = s[i];
        char second_char = s[i + 1];

        for (size_t j = i + 2; j < s.length() - 1; j++) {
            if (s[j] == first_char && s[j + 1] == second_char) {
                return true;
            }
        }
    }

    return false;
}

// Function that takes string and returns the score
int calculate_password_strength(const std::string& password) {
    // Reset the rules array
    reset_rules_array();

    // Check each rule and set the corresponding flag
    rules_activated[0] = rule1(password);
    rules_activated[1] = rule2(password);
    rules_activated[2] = rule3(password);
    rules_activated[3] = rule4(password);
    rules_activated[4] = rule5(password);
    rules_activated[5] = rule6(password);
    rules_activated[6] = rule7(password);

    // Calculate and return score
    int score = 0;
    for (int i = 0; i < NUMBER_OF_RULES; i++) {
        if (rules_activated[i]) {
            score++;
        }
    }

    return score;
}

int main() {
    int T;
    std::cin >> T;
    std::cin.ignore(); // To ignore the newline

    for (int i = 0; i < T; i++) {
        std::string password;
        std::getline(std::cin, password);
        std::cout << calculate_password_strength(password) << std::endl;
    }

    return 0;
}

File: Srednje skole/2026/Jacina sifre/solutions/main.py
import sys

RULE_1_LENGTH = 11

SPECIAL_RULE4 = set("!@#$%&*?_")
SPECIAL_RULE5 = set("!@#")


def rule1(s: str) -> bool:
    return len(s) >= RULE_1_LENGTH


def rule2(s: str) -> bool:
    return len(s) > 0 and ("a" <= s[0] <= "z")


def rule3(s: str) -> bool:
    return len(s) > 0 and ("0" <= s[-1] <= "9")


def rule4(s: str) -> bool:
    if len(s) < 2:
        return False
    return (s[1] in SPECIAL_RULE4) or (s[-2] in SPECIAL_RULE4)


def rule5(s: str) -> bool:
    has_lower = has_upper = has_digit = has_special = False

    for c in s:
        if "a" <= c <= "e":
            has_lower = True
        elif "A" <= c <= "E":
            has_upper = True
        elif "0" <= c <= "4":
            has_digit = True
        elif c in SPECIAL_RULE5:
            has_special = True

    count = has_lower + has_upper + has_digit + has_special
    return count >= 3


def rule6(s: str) -> bool:
    freq = {}
    for c in s:
        freq[c] = freq.get(c, 0) + 1
        if freq[c] >= 4:
            return True
    return False


def rule7(s: str) -> bool:
    # Non-overlapping repeated bigram:
    # exists i < j such that s[i:i+2] == s[j:j+2] and j - i >= 2
    if len(s) < 4:
        return False

    first_pos = {}  # bigram -> earliest start index
    for j in range(len(s) - 1):
        bg = s[j : j + 2]
        if bg in first_pos:
            if j - first_pos[bg] >= 2:
                return True
            # keep earliest index; it gives the best chance to reach >=2 later
        else:
            first_pos[bg] = j

    return False


def calculate_password_strength(password: str) -> int:
    checks = [
        rule1(password),
        rule2(password),
        rule3(password),
        rule4(password),
        rule5(password),
        rule6(password),
        rule7(password),
    ]
    return sum(checks)


def main() -> None:
    lines = sys.stdin.read().splitlines()
    if not lines:
        return
    t = int(lines[0].strip())
    out = []
    for i in range(1, 1 + t):
        pw = lines[i].strip()
        out.append(str(calculate_password_strength(pw)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()
